datalad_next/iter_collections/annexworktree.py

```diff
@@ -0,0 +1,123 @@
+    annexkey: str | None = None
+    annexsize: int | None = None
+    annexobjpath: PurePath | None = None
+
+
+@dataclass
+class AnnexWorktreeFileSystemItem(GitWorktreeFileSystemItem):
+    annexkey: str | None = None
+    annexsize: int | None = None
+
+
+def iter_annexworktree(
+        path: Path,
+        *,
+        untracked: str | None = 'all',
+        link_target: bool = False,
+        fp: bool = False,
+) -> Generator[AnnexWorktreeItem | AnnexWorktreeFileSystemItem, None, None]:
+
+    glsf = iter_gitworktree(
+        path,
+        untracked=untracked,
+        link_target=link_target,
+        fp=fp
+    )
+
+    git_fileinfo_store: list[Any] = list()
+    key_store: list[Any] = list()
+
+    with \
+            iter_subproc(
+                # we get the annex key for any filename (or empty if not annexed)
+                ['git', '-C', str(path), 'annex', 'find', '--anything', '--format=\${key}\n', '--batch'],
+                # intersperse items with newlines to trigger a batch run
+                # this avoids string operations to append newlines to items
+                input=intersperse(
+                    b'\n',
+                    # store all output of the git ls-find in the gitfileinfo
+                    # store
+                    route_out(
+                        glsf,
+                        git_fileinfo_store,
+                        lambda data: (str(data.name).encode(), [data])
+                    )
+                ),
+            ) as gaf, \
+            iter_subproc(
+                # get the key properties JSON-lines style
+                ['git', '-C', str(path), 'annex', 'examinekey', '--json', '--batch'],
+                # process all non-empty keys and store them in the key store,
+                # skip processing of empty keys and store an ignored value in
+                # the key store
+                input=route_out(
+                    itemize(gaf, sep=linesep_bytes, keep_ends=True),
+                    key_store,
+                    lambda data: (dont_process, [None])
+                                 if data == linesep_bytes
+                                 else (data, [data])
+                )
+            ) as gek:
+
+        for item in route_in(
+                route_in(
+                    load_json(itemize(gek, sep=linesep_bytes)),
+                    key_store,
+                    join_with_list,
+                ),
+                git_fileinfo_store,
+                join_with_list,
+        ):
+            yield AnnexWorktreeItem(
+                name=item[2].name,
+                gitsha=item[2].gitsha,
+                gittype=item[2].gittype,
+                annexkey=item[1].decode().strip() if item[1] else None,
+                annexsize=int(item[0]['bytesize']) if item[0] else None,
+                annexobjpath=PurePath(PurePosixPath(str(item[0]['objectpath'])))
+                             if item[0]
+                             else None,
```

Comment:
I think the semantics of the items in `item` deserve a comment. `item[0]` would be the `examinekey` record. `item[1]` is the keyname, which is also in `item[0]['key']`. `item[2]` is the gitworktree item.

With a different joiner (than `join_with_list`), this could clearer. I am under the impression that the joiner could produce `AnnexWorktreeItem` instances directly, rather than producing an intermediate container.

Reply:
> With a different joiner (than join_with_list), this could clearer. I am under the impression that the joiner could produce AnnexWorktreeItem instances directly, rather than producing an intermediate container.

That is a very good idea. It is applied in 1e1d3e57fab42ebd1b9b20c1e6e5d91ea69473ee